Use Instant to track session start time

martint opened the pull request:
This is needed for higher precision version of the special
datetime functions (localtimestamp, current_timestamp, etc),
since Instant can provide nanosecond precision vs the current
value in milliseconds.

findepi commented on presto-main/src/main/java/io/prestosql/testing/TestingConnectorSession.java:
```diff
@@ -48,7 +49,7 @@
     private final Locale locale;
     private final Optional<String> traceToken;
-    private final long startTime;
+    private final Instant startTime;
```
rename field and ctor to `start`

findepi commented on presto-main/src/main/java/io/prestosql/testing/TestingConnectorSession.java:
```diff
@@ -164,7 +165,7 @@ public static Builder builder()
         private final Locale locale = ENGLISH;
         private final Optional<String> traceToken = Optional.empty();
-        private Optional<Long> startTime = Optional.empty();
+        private Optional<Instant> startTime = Optional.empty();
```
`start`

findepi commented on presto-main/src/main/java/io/prestosql/testing/TestingConnectorSession.java:
```diff
@@ -181,7 +182,7 @@ public Builder setTimeZoneKey(TimeZoneKey timeZoneKey)
         }
 
-        public Builder setStartTime(long startTime)
+        public Builder setStartTime(Instant startTime)
```
`setStart`

findepi commented on presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -50,7 +52,7 @@ public void testFormatDateCannotImplicitlyAddTimeZoneToTimestampLiteral()
     {
         Session localSession = Session.builder(session)
-                .setStartTime(new DateTime(2017, 3, 1, 14, 30, 0, 0, DATE_TIME_ZONE).getMillis())
+                .setStartTime(Instant.ofEpochMilli(new DateTime(2017, 3, 1, 14, 30, 0, 0, DATE_TIME_ZONE).getMillis()))
```
In session builder you can leave an overload taking millis, this would improve readability of the code here.

martint replied:
We want to discourage usages where it's constructing the Instant from millis. Ideally, all these calls should use Java's datetime to instant conversions instead of using Joda's DateTime .

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorSession.java:
```diff
@@ -38,7 +39,15 @@ default String getUser()
+    /**
+     * @deprecated use {@link #getStart()} instead
+     */
+    default long getStartTime()
```
add `@Deprecated` too?

(should we have a static check?)

martint replied:
What do you mean by a static check?

electrum replied:
Using deprecated in the Javadoc without the annotation.

martint replied:
That makes sense. Is it even feasible to do that with checkstyle?

martint committed:
Use Instant to track session start time

This is needed for higher precision version of the special
datetime functions (localtimestamp, current_timestamp, etc),
since Instant can provide nanosecond precision vs the current
value in milliseconds.
